### PR TITLE
Add --installroot to YUM and DNF modules, issue #11310

### PR DESCRIPTION
**Pull request by berenddeschouwer**
This continues ansible-modules-core#1558, and
ansible-modules-core#1669

Allow specifying installroot for the yum and dnf modules
to install and remove packages in a location other than /.

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
modules/packaging/os/dnf.py

##### ANSIBLE VERSION
```
ansible 2.3.0 (devel 218bab80de) last updated 2017/01/04 14:48:31 (GMT +200)
```

##### SUMMARY
Add --installroot option for YUM and DNF

Allow for using YUM and DNF to install packages in a directory other
than /.

I'm continuing other people's work, but I use it for LTSP and other root and jail-ed installs.  YUM --installroot was submitted before as #11310, and ansible-core-modules (since included in ansible) as issues # 1558 and # 1669.  I also included it in DNF, and kept the previous change author in the yum module.

Tests are included.

### CODE REVIEW

File: lib/ansible/modules/packaging/os/dnf.py [hunk 1]
```diff
@@ -86,6 +86,15 @@
...
     aliases: []
 
+  installroot:
+    description:
+      - Specifies an alternative installroot, relative to which all packages
+        will be installed.
+    required: false
+    version_added: "2.3"
+    default: "/"
+    aliases: []
```
Comment (gundalow) [inline lib/ansible/modules/packaging/os/dnf.py]: As this is empty you can remove this.

Could you please remove them from the other docs above.

File: lib/ansible/modules/packaging/os/dnf.py [hunk 2]
```diff
@@ -189,6 +199,12 @@ def _configure_base(module, base, conf_file, disable_gpg_check):
     # Don't prompt for user confirmations
     conf.assumeyes = True
 
+    # Set installroot
+    if installroot and installroot != '/':
```
Comment (gundalow) [inline lib/ansible/modules/packaging/os/dnf.py]: Should the default in `argument_spec` (and the docs) just be set to `/`? then this can be simplified?

**Review by gundalow — CHANGES_REQUESTED** [review]
needs_revision

Docs tidy up, and a suggestion on how to simplify the logic.

Bonus points for tests
Extra bonus points for mktemp

File: lib/ansible/modules/packaging/os/dnf.py (hunk 1 above)
Comment (berenddeschouwer) [reply lib/ansible/modules/packaging/os/dnf.py]: I've made this change

File: lib/ansible/modules/packaging/os/dnf.py (hunk 2 above)
Comment (berenddeschouwer) [reply lib/ansible/modules/packaging/os/dnf.py]: ... and this change

**Comment by abadger** [comment]
Merged to devel for the 2.3 release.  Thanks!

**Comment by mattclay** [comment]
Tests are failing on centos 6 and 7 with these changes: https://app.shippable.com/runs/586dbb659a67460f0094e261/9/tests

**Comment by mattclay** [comment]
The tests passed on the PR due to an unrelated bug that was present at the time the tests were run. The bug caused the tests to only partially run, skipping the portions that would have failed.

**Comment by mattclay** [comment]
@berenddeschouwer, @abadger: I've reverted this PR with #19924 to restore tests to a functional state. These changes (with fixes) will need to be re-submitted in a new PR.

**Comment by mattclay** [comment]
I tried fixing the tests by changing `ansible_lsb.major_release` to `ansible_distribution_major_version` instead. That resolved the first failure, but then the tests failed farther in.

**Comment by berenddeschouwer** [comment]
Could you tell me how to duplicate the fault?  I understand that something causes my test machine to set ansible_lsb (gather-facts?) but not on shippable.com.

I just did
- source hacking/env_setup
- TEST_FLAGS="--tags test_yum" make destructive

Is there something else I should be running?

**Comment by gundalow** [comment]
`sudo $ ./test/runner/ansible-test integration --docker centos6 yum`

Where the list of distros is defined at https://github.com/ansible/ansible/blob/devel/shippable.yml#L10

**Comment by abadger** [comment]
I think I've got a fix: https://github.com/ansible/ansible/pull/19925  It's pretty late here, though, so I'm going to go to sleep soon.  feel free to build off of that if you see something else that needs changing.

**Pull request merged**